File: SeriesGuide/src/main/java/com/battlelancer/seriesguide/adapters/TabStripAdapter.java
package com.battlelancer.seriesguide.adapters;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.StringRes;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.ViewPager;
import com.battlelancer.seriesguide.R;
import com.battlelancer.seriesguide.widgets.SlidingTabLayout;
import java.util.ArrayList;
import java.util.Locale;

/**
 * Helper class for easy setup of a {@link com.battlelancer.seriesguide.widgets.SlidingTabLayout}.
 */
public class TabStripAdapter extends FragmentPagerAdapter {

    static final class TabInfo {

        private final Class<?> fragmentClass;
        private final Bundle args;
        private final int titleRes;

        TabInfo(Class<?> fragmentClass, Bundle args, int titleRes) {
            this.fragmentClass = fragmentClass;
            this.args = args;
            this.titleRes = titleRes;
        }
    }

    private final ArrayList<TabInfo> tabs = new ArrayList<>();
    private final Context context;
    private final FragmentManager fragmentManager;

    private final ViewPager viewPager;
    private final SlidingTabLayout tabLayout;

    public TabStripAdapter(FragmentManager fragmentManager, Context context, ViewPager pager,
            SlidingTabLayout tabLayout) {
        super(fragmentManager);
        this.fragmentManager = fragmentManager;
        this.context = context;

        // setup view pager
        viewPager = pager;
        viewPager.setAdapter(this);

        // setup tabs
        this.tabLayout = tabLayout;
        this.tabLayout.setCustomTabView(R.layout.tabstrip_item_allcaps, R.id.textViewTabStripItem);
        this.tabLayout.setSelectedIndicatorColors(ContextCompat.getColor(context, R.color.white));
        this.tabLayout.setViewPager(viewPager);
    }

    /**
     * Adds a new tab. Make sure to call {@link #notifyTabsChanged} after you have added them all.
     */
    public void addTab(@StringRes int titleRes, Class<?> fragmentClass, Bundle args) {
        tabs.add(new TabInfo(fragmentClass, args, titleRes));
    }

    /**
     * Update an existing tab. Make sure to call {@link #notifyTabsChanged} afterwards.
     */
    public void updateTab(int titleRes, Class<?> fragmentClass, Bundle args, int position) {
        if (position >= 0 && position < tabs.size()) {
            // update tab info
            tabs.set(position, new TabInfo(fragmentClass, args, titleRes));

            // find current fragment of tab
            Fragment oldFragment = fragmentManager
                    .findFragmentByTag(makeFragmentName(viewPager.getId(), getItemId(position)));
            // remove it
            FragmentTransaction transaction = fragmentManager.beginTransaction();
            transaction.remove(oldFragment);
            transaction.commitAllowingStateLoss();
            fragmentManager.executePendingTransactions();
        }
    }

    /**
     * Notifies the adapter and tab strip that the tabs have changed.
     */
    public void notifyTabsChanged() {
        notifyDataSetChanged();
        tabLayout.setViewPager(viewPager);
    }

    @Override
    public Fragment getItem(int position) {
        TabInfo tab = tabs.get(position);
        return Fragment.instantiate(context, tab.fragmentClass.getName(), tab.args);
    }

    @Override
    public int getCount() {
        return tabs.size();
    }

    @Override
    public CharSequence getPageTitle(int position) {
        TabInfo tabInfo = tabs.get(position);
        if (tabInfo != null) {
            return context.getString(tabInfo.titleRes).toUpperCase(Locale.getDefault());
        }
        return "";
    }

    /**
     * Copied from FragmentPagerAdapter.
     */
    private static String makeFragmentName(int viewId, long id) {
        return "android:switcher:" + viewId + ":" + id;
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/ui/TabStripAdapter.kt
package com.battlelancer.seriesguide.ui

import android.annotation.SuppressLint
import android.os.Bundle
import androidx.annotation.StringRes
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import com.battlelancer.seriesguide.util.ThemeUtils.setDefaultStyle
import com.uwetrottmann.seriesguide.widgets.SlidingTabLayout

/**
 * Helper class for easy setup of a {@link SlidingTabLayout} with a mutable set of tabs.
 * Requires that tabs each have a unique title string resource
 * as it is used to uniquely identify a tab.
 */
open class TabStripAdapter(
    private val fragmentActivity: FragmentActivity,
    private val viewPager: ViewPager2,
    private val tabLayout: SlidingTabLayout
) : FragmentStateAdapter(fragmentActivity) {

    private val tabs = ArrayList<TabInfo>()

    private val tabTitleSupplier = SlidingTabLayout.TabTitleSupplier { position ->
        val tabInfo = tabs.getOrNull(position)
        return@TabTitleSupplier if (tabInfo != null) {
            tabLayout.context.getString(tabInfo.titleRes)
        } else ""
    }

    init {
        // Preload next/previous page for smoother swiping.
        viewPager.offscreenPageLimit = 1
        viewPager.adapter = this

        tabLayout.setDefaultStyle()
        tabLayout.setViewPager2(viewPager, tabTitleSupplier)
    }

    /**
     * Adds a new tab at the end.
     *
     * Make sure to call [notifyTabsChanged] after you have added them all.
     */
    fun addTab(@StringRes titleRes: Int, fragmentClass: Class<*>, args: Bundle?) {
        tabs.add(tabs.size, TabInfo(fragmentClass, args, titleRes))
    }

    /**
     * Adds a new tab at the given position.
     *
     * Make sure to call [notifyTabsChanged] after you have added them all.
     */
    fun addTab(@StringRes titleRes: Int, fragmentClass: Class<*>, args: Bundle?, position: Int) {
        tabs.add(position, TabInfo(fragmentClass, args, titleRes))
    }

    /**
     * Notifies the adapter and tab strip that the tabs have changed.
     */
    @SuppressLint("NotifyDataSetChanged")
    fun notifyTabsChanged() {
        notifyDataSetChanged()
        // update tabs
        tabLayout.setViewPager2(viewPager, tabTitleSupplier)
    }

    // Using titleRes instead of introducing an ID property as it is unique for tabs of this app.
    override fun getItemId(position: Int): Long {
        return if (position < tabs.size) {
            tabs[position].titleRes.toLong()
        } else {
            RecyclerView.NO_ID
        }
    }

    override fun containsItem(itemId: Long): Boolean =
        tabs.find { it.titleRes.toLong() == itemId } != null

    override fun getItemCount(): Int = tabs.size

    override fun createFragment(position: Int): Fragment {
        val tab = tabs[position]
        return fragmentActivity.supportFragmentManager.fragmentFactory.instantiate(
            fragmentActivity.classLoader,
            tab.fragmentClass.name
        ).apply { tab.args?.let { arguments = it } }
    }

    data class TabInfo(
        val fragmentClass: Class<*>,
        val args: Bundle?,
        val titleRes: Int
    )
}